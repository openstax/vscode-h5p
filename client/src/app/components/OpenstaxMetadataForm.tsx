import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMinusCircle, faPlusCircle } from '@fortawesome/free-solid-svg-icons';

const TypedInputSet = ({
  title,
  inputs,
  handleAddInput,
  handleRemoveInput,
  handleInputChange,
}) => {
  return (
    <div>
      <div className="container">
        <div className="row">
          <div className="col-8">
            <h3>{title}</h3>
          </div>
          <div className="col-4 pt-2">
            <FontAwesomeIcon
              icon={faPlusCircle}
              onClick={() => handleAddInput()}
              style={{ cursor: 'pointer' }}
            />
          </div>
        </div>
        {inputs.map((input, index) => (
          <div key={index} className="row pt-1 pb-1">
            <div className="col-8">
              <input
                value={input.value}
                onChange={(event) =>
                  handleInputChange(index, event.target.value)
                }
                style={{ width: '100%' }}
              />
            </div>
            <div className="col-4">
              <FontAwesomeIcon
                icon={faMinusCircle}
                onClick={() => handleRemoveInput(index)}
                style={{ cursor: 'pointer' }}
              />
            </div>
          </div>
        ))}
      </div>
    </div>
  );
};

export default class OpenstaxMetadataForm extends React.Component {
  state = {
    book: [],
    lo: [],
  };

  save(contentId) {
    console.log('Save', contentId);
  }

  render() {
    const handleInputChange = (type, index, value) => {
      const inputs = this.state[type];
      const newInputs = [...inputs];
      newInputs[index].value = value;
      this.setState({ [type]: newInputs });
    };

    const handleAddInput = (type) => {
      const inputs = this.state[type];
      this.setState({ [type]: [...inputs, { value: '' }] });
    };

    const handleRemoveInput = (type, index) => {
      const inputs = this.state[type];
      const newInputs = [...inputs];
      newInputs.splice(index, 1);
      this.setState({ [type]: newInputs });
    };

    const inputHandlerProps = (type) => {
      return {
        inputs: this.state[type],
        handleAddInput: () => handleAddInput(type),
        handleRemoveInput: (index) => handleRemoveInput(type, index),
        handleInputChange: (index, value) =>
          handleInputChange(type, index, value),
      };
    };

    return (
      <div className="container">
        <div className="row">
          <div className="col-6">
            <TypedInputSet title={'Book'} {...inputHandlerProps('book')} />
          </div>
          <div className="col-6">
            <TypedInputSet title={'LO'} {...inputHandlerProps('lo')} />
          </div>
        </div>
        <div>Testing 1 2 3 state.saving</div>
      </div>
    );
  }
}
