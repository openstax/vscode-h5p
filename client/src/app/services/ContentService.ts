import type {
  IEditorModel,
  IPlayerModel,
  IContentMetadata,
} from '@lumieducation/h5p-server';
import axios from 'axios';

export interface IContentListEntry {
  h5PUrl: string;
  contentId: string;
  mainLibrary: string;
  title: string;
  originalNewKey?: string;
}

export interface IContentService {
  delete(contentId: string): Promise<void>;
  getEdit(contentId: string): Promise<IEditorModel>;
  getPlay(contentId: string): Promise<IPlayerModel>;
  list(): Promise<IContentListEntry[]>;
  save(
    contentId: string,
    requestBody: { library: string; params: any }
  ): Promise<{ contentId: string; metadata: IContentMetadata }>;
  generateDownloadLink(contentId: string): string;
}

export class ContentService implements IContentService {
  /**
   *
   */
  constructor(protected baseUrl: string = '') {}

  private csrfToken: string | undefined = undefined;

  delete = async (contentId: string): Promise<void> => {
    console.log(`ContentService: deleting ${contentId}...`);
    try {
      const result = await axios.delete(`${this.baseUrl}/${contentId}`, {
        method: 'delete',
        headers: {
          'CSRF-Token': this.csrfToken ?? '',
        },
      });
    } catch (err) {
      console.error(err);
      throw err;
    }
  };

  getEdit = async (contentId: string): Promise<IEditorModel> => {
    console.log(`ContentService: Getting information to edit ${contentId}...`);
    try {
      const res = await axios.get(`${this.baseUrl}/${contentId}/edit`);
      return res.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  };

  getPlay = async (contentId: string): Promise<IPlayerModel> => {
    console.log(`ContentService: Getting information to play ${contentId}...`);
    try {
      const res = await axios.get(`${this.baseUrl}/${contentId}/play`);
      return res.data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  };

  list = async (): Promise<IContentListEntry[]> => {
    console.log(`ContentService: Listing content objects`);
    try {
      const result = await axios.get(this.baseUrl);
      return result.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  };

  save = async (
    contentId: string,
    requestBody: { library: string; params: any }
  ): Promise<{ contentId: string; metadata: IContentMetadata }> => {
    if (contentId) {
      console.log(`ContentService: Saving new content.`);
    } else {
      console.log(`ContentService: Saving content ${contentId}`);
    }

    const body = JSON.stringify(requestBody);
    try {
      const res = contentId
        ? await axios.patch(`${this.baseUrl}/${contentId}`, body, {
            headers: {
              'Content-Type': 'application/json',
              'CSRF-Token': this.csrfToken ?? '',
            },
          })
        : await axios.post(this.baseUrl, body, {
            headers: {
              'Content-Type': 'application/json',
              'CSRF-Token': this.csrfToken ?? '',
            },
          });
      return res.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  };
  generateDownloadLink = (contentId: string): string =>
    `${this.baseUrl}/download/${contentId}`;

  setCsrfToken = (csrfToken): void => {
    this.csrfToken = csrfToken;
  };
  getCsrfToken = (): string | undefined => {
    return this.csrfToken;
  };
}
