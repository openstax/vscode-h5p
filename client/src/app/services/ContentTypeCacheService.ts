import axios from 'axios';
/**
 * This service performs queries at the REST endpoint of the content type cache.
 */
export default class ContentTypeCacheService {
  constructor(private baseUrl: string) {}

  /**
   * Gets the last update date and time.
   */
  public async getCacheUpdate(): Promise<Date | null> {
    try {
      const response = await axios.get(`${this.baseUrl}/update`);

      const { lastUpdate } = response.data;
      return lastUpdate === null ? null : new Date(lastUpdate);
    } catch (err) {
      console.error(err);
      throw err;
    }
  }

  /**
   * Triggers a content type cache update that will contact the H5P Hub and
   * retrieve the latest content type list.
   */
  public async postUpdateCache(): Promise<Date> {
    try {
      const response = await axios.post(`${this.baseUrl}/update`);
      return new Date(response.data.lastUpdate);
    } catch (err) {
      console.error(err);
      throw err;
    }
  }
}
