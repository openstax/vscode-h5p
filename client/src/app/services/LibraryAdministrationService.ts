import type {
  IInstalledLibrary,
  ILibraryAdministrationOverviewItem,
} from '@lumieducation/h5p-server';
import axios from 'axios';

/**
 * The data model used to display the library list.
 */
export interface ILibraryViewModel extends ILibraryAdministrationOverviewItem {
  details?: IInstalledLibrary & {
    dependentsCount: number;
    instancesAsDependencyCount: number;
    instancesCount: number;
    isAddon: boolean;
  };
  isDeleting?: boolean;
  isShowingDetails?: boolean;
}

/**
 *
 */
export class LibraryAdministrationService {
  constructor(private baseUrl: string) {}

  public async deleteLibrary(library: ILibraryViewModel): Promise<void> {
    try {
      const response = await axios.delete(
        `${this.baseUrl}/${library.machineName}-${library.majorVersion}.${library.minorVersion}`,
      );

      return response.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  }

  public async getLibraries(): Promise<ILibraryViewModel[]> {
    try {
      const response = await axios.get(this.baseUrl);

      return response.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  }

  public async getLibrary(library: ILibraryViewModel): Promise<
    IInstalledLibrary & {
      dependentsCount: number;
      instancesAsDependencyCount: number;
      instancesCount: number;
      isAddon: boolean;
    }
  > {
    try {
      const response = await axios.get(
        `${this.baseUrl}/${library.machineName}-${library.majorVersion}.${library.minorVersion}`,
      );
      return response.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  }

  public async patchLibrary(
    library: ILibraryViewModel,
    changes: Partial<ILibraryViewModel>,
  ): Promise<ILibraryViewModel> {
    try {
      const response = await axios.patch(
        `${this.baseUrl}/${library.machineName}-${library.majorVersion}.${library.minorVersion}`,
        {
          headers: {
            'Content-Type': 'application/json;charset=UTF-8',
          },
          body: JSON.stringify(changes),
        },
      );
      return response.data;
    } catch (err) {
      console.error(err);
      throw err;
    }
  }

  public async postPackage(
    file: File,
  ): Promise<{ installed: number; updated: number }> {
    const formData = new FormData();
    formData.append('file', file);
    try {
      const response = await axios.post(this.baseUrl, formData);
      const result = response.data;
      return { installed: result.installed, updated: result.updated };
    } catch (err) {
      console.error(err);
      throw err;
    }
  }
}
