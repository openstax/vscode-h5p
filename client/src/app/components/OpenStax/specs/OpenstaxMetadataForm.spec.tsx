import {
  cleanup,
  fireEvent,
  render,
  getByDisplayValue,
} from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import React from 'react';

import OpenstaxMetadataForm from '../OpenstaxMetadataForm';
import { IContentService } from '../../../services/ContentService';
import {
  AP_BOOKS,
  AP_HISTORY_BOOKS,
  AP_SCIENCE_BOOKS,
  NURSING_BOOKS,
} from '../constants';

const SEL_INPUT_SET_ADD = '[data-control-type="input-set-add"]';
const SEL_INPUT_SET_REM = '[data-control-type="input-set-subtract"]';

const SEL_BOOK = '[data-control-type="book"]';
const SEL_BOOK_REM = '[data-control-type="remove-book"]';
const SEL_BOOK_ADD = '[data-control-type="add-book"]';

describe('OpenstaxMetadataForm', () => {
  const defaultMockContentService = {
    getOSMeta: jest.fn().mockResolvedValue({}),
    saveOSMeta: jest.fn().mockResolvedValue(undefined),
  } as any as IContentService;

  const defaultFormProps = {
    contentId: 'new',
    contentService: defaultMockContentService,
    onSaveError: jest.fn(),
  };

  const minFormData = {
    nickname: 'test',
    blooms: 1,
  };

  const createForm = async (formProps) => {
    const openstaxForm = React.createRef<OpenstaxMetadataForm>();
    let all;
    await act(async () => {
      all = render(<OpenstaxMetadataForm {...formProps} ref={openstaxForm} />);
    });
    return { openstaxForm, ...all };
  };

  const revealForm = async ({ getByText }) => {
    const formTitle = await getByText('OpenStax Metadata');
    const formParent = formTitle.parentElement;
    expect(formTitle).toBeTruthy();
    expect(formParent).toBeTruthy();
    // Open the Accordion so that its children render
    fireEvent.click(formTitle.parentElement!, { button: 1 });
  };

  const initForm = async (formData, formPropsOverride) => {
    const formProps = {
      ...defaultFormProps,
      contentService: {
        ...defaultMockContentService,
        getOSMeta: jest.fn().mockResolvedValue(formData),
      },
      ...formPropsOverride,
    };
    const controller = await createForm(formProps);
    await revealForm(controller);
    return controller;
  };

  const initFormWithMinData = async (args: {
    formDataOverride?;
    formPropsOverride?;
  }) => {
    const { formDataOverride = {}, formPropsOverride = {} } = args;
    return await initForm(
      { ...minFormData, ...formDataOverride },
      formPropsOverride
    );
  };

  const selectOption = (container: HTMLElement, idx: number, value: string) => {
    const input = [
      ...container.querySelectorAll('input[role="combobox"]'),
    ].find((el, i) => el.id.match(/^react-select-\d+-input$/) && i === idx);
    if (input === undefined) {
      throw new Error('Failed to find input for select box');
    }
    // Type the value then press Enter
    act(() => {
      fireEvent.change(input, { target: { value } });
    });
    act(() => {
      fireEvent.keyDown(input, { keyCode: 13 });
    });
  };

  afterEach(cleanup);

  beforeEach(() => jest.resetAllMocks());

  describe('Default Input Set', () => {
    it('is rendered', async () => {
      const controller = await createForm(defaultFormProps);
      const { getByText } = controller;
      await revealForm(controller);
      expect(await getByText('Books')).toBeTruthy();
    });
    it('renders when there is an error in getOSMeta', async () => {
      const { getByText } = await initFormWithMinData({
        formPropsOverride: {
          contentService: {
            ...defaultMockContentService,
            getOSMeta: jest.fn().mockRejectedValue(new Error('Test')),
          },
        },
      });
      expect(await getByText('Books')).toBeTruthy();
    });
    [-1, 1].forEach((inc) => {
      const isAdd = inc > 0;
      it(`can ${isAdd ? 'add' : 'remove'} inputs in a set`, async () => {
        const { container } = await initFormWithMinData({
          formDataOverride: {
            'module-id': [{ module: 'a/b/c', 'element-id': '' }],
          },
        });
        const inputCountBefore = container.querySelectorAll('input').length;
        const expectedCount = inputCountBefore + inc;
        const selector = isAdd ? SEL_INPUT_SET_ADD : SEL_INPUT_SET_REM;
        const button = container.querySelector(selector)?.firstElementChild;
        expect(button).not.toBeFalsy();
        act(() => {
          fireEvent.click(button!, { button: 1 });
        });
        expect(container.querySelectorAll('input').length).toBe(expectedCount);
      });
    });
  });
  describe('Conditional Input Set', () => {
    const tests: Array<[string, string, string[]]> = [
      ['AP LO', 'AP Learning Objectives', AP_BOOKS],
      ['Science Practice', 'Science Practice', AP_SCIENCE_BOOKS],
      ['Historical Thinking', 'Historical Thinking', AP_HISTORY_BOOKS],
      ['AACN', 'AACN', NURSING_BOOKS],
      ['NCLEX', 'NCLEX', NURSING_BOOKS],
    ];
    tests.forEach(([name, text, books]) => {
      books.forEach((book) => {
        it(`shows "${name}" when "${book}" is selected`, async () => {
          const { getByText } = await initFormWithMinData({
            formDataOverride: { books: { [book]: {} } },
          });
          expect(await getByText(text)).toBeTruthy();
        });
      });
    });
  });
  describe('Book inputs', () => {
    const getBook = (
      container: HTMLElement,
      idx: number
    ): HTMLElement | undefined => {
      const el = container.querySelectorAll(SEL_BOOK)[idx];
      return el === undefined ? undefined : (el as HTMLElement);
    };
    const getBookInput = (
      container: HTMLElement,
      bookIdx: number,
      name: string
    ): HTMLElement | undefined => {
      const el = Array.from(
        Array.from(container.querySelectorAll(SEL_BOOK))[
          bookIdx
        ]?.querySelectorAll('.container') ?? []
      ).find((el) => el.querySelector('h3')?.textContent === name);
      return el === undefined ? undefined : (el as HTMLElement);
    };
    [-1, 1].forEach((inc) => {
      const isAdd = inc > 0;
      it(`can ${isAdd ? 'add' : 'remove'} books`, async () => {
        const { container } = await initFormWithMinData({
          formDataOverride: {
            books: {
              'stax-psy': {
                lo: ['1-2-3'],
              },
            },
          },
        });
        const bookCountBefore = container.querySelectorAll(SEL_BOOK).length;
        const expectedCount = bookCountBefore + inc;
        const selector = isAdd ? SEL_BOOK_ADD : SEL_BOOK_REM;
        const button = container.querySelector(selector)?.firstElementChild;
        expect(button).toBeTruthy();
        act(() => {
          fireEvent.click(button!, { button: 1 });
        });
        expect(container.querySelectorAll(SEL_BOOK).length).toBe(expectedCount);
      });
    });
    it('correctly adds and removes inputs in a set', async () => {
      const { container } = await initFormWithMinData({
        formDataOverride: {
          books: {
            'stax-psy': {
              lo: ['1-2-3'],
            },
          },
        },
      });
      const bookInput = getBookInput(container, 0, 'Learning Objectives')!;
      const addButton =
        bookInput.querySelector(SEL_INPUT_SET_ADD)?.firstElementChild;
      const remButton =
        bookInput.querySelector(SEL_INPUT_SET_REM)?.firstElementChild;

      expect(bookInput.querySelectorAll(SEL_INPUT_SET_REM).length).toBe(1);

      act(() => {
        fireEvent.click(addButton!, { button: 1 });
      });
      expect(bookInput.querySelectorAll(SEL_INPUT_SET_REM).length).toBe(2);

      act(() => {
        fireEvent.click(remButton!, { button: 1 });
      });
      expect(bookInput.querySelectorAll(SEL_INPUT_SET_REM).length).toBe(1);
    });
    it('correctly modifies inputs in a set', async () => {
      let count = 0;
      const getValue = () => {
        return `not-modified-${count++}`;
      };
      const oldValue = getValue();
      const newValue = 'stax-psy.lo[1]';
      const { container, openstaxForm } = await initFormWithMinData({
        formDataOverride: {
          books: {
            'stax-should-be-removed': {
              lo: [getValue()],
            },
            'stax-psy': {
              lo: [getValue(), oldValue, getValue()],
            },
            'stax-anything': {
              lo: [getValue()],
            },
          },
        },
      });
      act(() => {
        fireEvent.change(
          getByDisplayValue(
            getBookInput(container, 1, 'Learning Objectives')!,
            oldValue
          )!,
          {
            target: { value: newValue },
          }
        );
      });
      // Remove a book
      act(() => {
        fireEvent.click(
          container.querySelector(SEL_BOOK_REM)!.firstElementChild,
          { button: 1 }
        );
      });
      // Add a book
      act(() => {
        fireEvent.click(
          container.querySelector(SEL_BOOK_ADD)!.firstElementChild,
          { button: 1 }
        );
      });
      // Add another option
      selectOption(getBook(container, 2)!, 0, 'stax-amfg');
      expect(openstaxForm.current!.encodedValues).toMatchSnapshot();
    });
    it('correctly modifies inputs', async () => {
      let count = 0;
      const getValue = () => {
        return `not-modified-${count++}`;
      };
      const oldValue = '';
      const newValue = 'modified';
      const { container, openstaxForm } = await initFormWithMinData({
        formDataOverride: {
          books: {
            'stax-nutrition': {
              aacn: getValue(),
            },
            'stax-pharmacology': {
              // Empty to test adding state for inputs
            },
            'stax-pophealth': {
              aacn: getValue(),
            },
          },
        },
      });
      act(() => {
        fireEvent.change(
          getByDisplayValue(getBookInput(container, 1, 'AACN')!, oldValue)!,
          {
            target: { value: newValue },
          }
        );
      });
      expect(openstaxForm.current!.encodedValues).toMatchSnapshot();
    });
    describe('handleBookChange', () => {
      it('keeps input values when they exist on the new book', async () => {
        const { container, openstaxForm } = await initFormWithMinData({
          formDataOverride: {
            books: {
              'stax-amfg': {
                lo: ['test-carry-state-to-stax-usgovt'],
              },
              'stax-psy': {
                lo: ['test-keep-state-on-stax-psy'],
              },
            },
          },
        });
        selectOption(getBook(container, 0)!, 0, 'stax-usgovt');
        expect(openstaxForm.current!.encodedValues).toMatchSnapshot();
      });
    });
  });
  describe('encode/decode form state', () => {
    const formDataEncoded = {
      ...minFormData,
      'module-id': ['m000001', 'm000002', 'm000003#term-03'].map((id) => {
        const splitValue = id.split('#');
        return {
          module: `modules/${splitValue[0]}/index.cnxml`,
          'element-id': splitValue[1] ?? '',
        };
      }),
      books: {
        'stax-psy': {
          lo: ['00-00-01'],
        },
        'stax-apbio': {
          'ap-lo': ['ABC.0.F'],
          'science-practice': 'concept-explanation',
        },
      },
    };
    it('decodes form state when loading and encodes when saving', async () => {
      const controller = await initFormWithMinData({
        formDataOverride: formDataEncoded,
      });
      const { openstaxForm } = controller;
      await openstaxForm.current!.save();
      expect(
        openstaxForm.current!.decodeValues(formDataEncoded)
      ).toMatchSnapshot();
      expect(
        (defaultMockContentService.saveOSMeta as jest.Mock).mock.calls
      ).toMatchSnapshot();
    });
  });
  describe('Form validation', () => {
    it('stores information about field validity', async () => {
      const nicknameInitialValue = 'supercoolnickname';
      const errorValue = nicknameInitialValue + '+';
      const { openstaxForm, getByDisplayValue } = await initFormWithMinData({
        formDataOverride: { nickname: nicknameInitialValue },
      });
      const nickInput = await getByDisplayValue(nicknameInitialValue);
      expect(nickInput).toBeTruthy();
      // Validity checks happen in handleInputChange
      act(() => {
        fireEvent.change(nickInput, { target: { value: errorValue } });
      });
      await openstaxForm.current!.save();
      expect(defaultMockContentService.saveOSMeta).not.toHaveBeenCalled();
      expect(nickInput.value).toBe(errorValue);
      expect(openstaxForm.current!.isInputValid).toBe(false);
      expect(defaultFormProps.onSaveError).toHaveBeenCalledWith(
        'OpenStax Metadata: Value for "nickname" is invalid'
      );
    });
    it('rejects invalid book input set values', async () => {
      const { openstaxForm: openstaxFormBadLO, getByDisplayValue } =
        await initFormWithMinData({
          formDataOverride: {
            books: {
              'stax-phys': {
                lo: ['test bookInputSetHandlerProps.getIdx', 'test-lo-field'],
              },
            },
          },
        });
      const loInput = getByDisplayValue('test-lo-field');
      expect(loInput).toBeTruthy();
      act(() => {
        fireEvent.change(loInput!, { target: { value: 'invalid ' } });
      });
      await openstaxFormBadLO.current!.save('new');
      expect(defaultFormProps.onSaveError).toHaveBeenCalledWith(
        'OpenStax Metadata: Value for "lo" is invalid'
      );
      expect(defaultFormProps.contentService.saveOSMeta).not.toBeCalled();
    });
    it('rejects invalid book input values', async () => {
      const { openstaxForm: openstaxFormBadLO, getByDisplayValue } =
        await initFormWithMinData({
          formDataOverride: {
            books: {
              'stax-nutrition': {
                aacn: 'test-aacn',
              },
            },
          },
        });
      const loInput = getByDisplayValue('test-aacn');
      expect(loInput).toBeTruthy();
      act(() => {
        fireEvent.change(loInput!, { target: { value: 'invalid ' } });
      });
      await openstaxFormBadLO.current!.save('new');
      expect(defaultFormProps.onSaveError).toHaveBeenCalledWith(
        'OpenStax Metadata: Value for "aacn" is invalid'
      );
      expect(defaultFormProps.contentService.saveOSMeta).not.toBeCalled();
    });
    it('does not save when information is missing', async () => {
      const formProps = { ...defaultFormProps };
      const { openstaxForm } = await createForm(formProps);
      await openstaxForm.current!.save('new');
      expect(formProps.onSaveError.mock.calls).toMatchSnapshot();
      expect(formProps.contentService.saveOSMeta).not.toBeCalled();

      (formProps.contentService.saveOSMeta as jest.Mock).mockReset();
      formProps.onSaveError.mockReset();
      cleanup();

      const { openstaxForm: openstaxFormNoNick } = await initFormWithMinData({
        formDataOverride: { nickname: '' },
      });
      await openstaxFormNoNick.current!.save('new');
      expect(formProps.onSaveError.mock.calls).toMatchSnapshot();
      expect(formProps.contentService.saveOSMeta).not.toBeCalled();
    });
    it('does save when all required fields are present', async () => {
      const formProps = {
        ...defaultFormProps,
        contentService: {
          ...defaultMockContentService,
          getOSMeta: jest.fn().mockResolvedValue(minFormData),
        } as any as IContentService,
      };
      const { openstaxForm } = await createForm(formProps);
      expect(openstaxForm.current).toBeTruthy();
      await openstaxForm.current!.save('does not matter');
      expect(formProps.contentService.getOSMeta).toBeCalledTimes(1);
      expect(formProps.onSaveError).not.toBeCalled();
      expect(formProps.contentService.saveOSMeta).toBeCalledTimes(1);
    });
  });
});
