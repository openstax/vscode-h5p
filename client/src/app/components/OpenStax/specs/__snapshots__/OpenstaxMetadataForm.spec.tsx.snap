// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenstaxMetadataForm Book inputs correctly modifies inputs 1`] = `
{
  "assignment_type": null,
  "blooms": null,
  "books": [
    {
      "aacn": "not-modified-0",
      "name": "stax-nutrition",
    },
    {
      "aacn": "modified",
      "name": "stax-pharmacology",
    },
    {
      "aacn": "not-modified-1",
      "name": "stax-pophealth",
    },
  ],
  "collaborator_solutions": [],
  "dok": null,
  "errata_id": "test",
  "feature_id": null,
  "feature_page": null,
  "is_solution_public": true,
  "nickname": "new",
  "time": null,
}
`;

exports[`OpenstaxMetadataForm Book inputs correctly modifies inputs in a set 1`] = `
{
  "assignment_type": null,
  "blooms": null,
  "books": [
    {
      "lo": [
        "not-modified-2",
        "stax-psy.lo[1]",
        "not-modified-3",
      ],
      "name": "stax-psy",
    },
    {
      "lo": [
        "not-modified-4",
      ],
      "name": "stax-anything",
    },
  ],
  "collaborator_solutions": [],
  "dok": null,
  "errata_id": "test",
  "feature_id": null,
  "feature_page": null,
  "is_solution_public": true,
  "nickname": "new",
  "time": null,
}
`;

exports[`OpenstaxMetadataForm Book inputs handleBookChange keeps input values when they exist on the new book 1`] = `
{
  "assignment_type": null,
  "blooms": null,
  "books": [
    {
      "lo": [
        "test-carry-state-to-stax-usgovt",
      ],
      "name": "stax-usgovt",
    },
    {
      "lo": [
        "test-keep-state-on-stax-psy",
      ],
      "name": "stax-psy",
    },
  ],
  "collaborator_solutions": [],
  "dok": null,
  "errata_id": "test",
  "feature_id": null,
  "feature_page": null,
  "is_solution_public": true,
  "nickname": "new",
  "time": null,
}
`;

exports[`OpenstaxMetadataForm Form validation does not save when information is missing 1`] = `[]`;

exports[`OpenstaxMetadataForm Form validation does not save when information is missing 2`] = `
[
  [
    "OpenStax Metadata: Value for "nickname" is invalid",
  ],
]
`;

exports[`OpenstaxMetadataForm encode/decode form state decodes form state when loading and encodes when saving 1`] = `
{
  "aacn": [],
  "aplo": [],
  "books": [],
  "errata_id": {
    "isValid": true,
    "value": "test",
  },
  "hts": [],
  "is_solution_public": {
    "isValid": true,
    "value": "true",
  },
  "lo": [],
  "nclex": [],
  "rp": [],
  "science_practice": [],
}
`;

exports[`OpenstaxMetadataForm encode/decode form state decodes form state when loading and encodes when saving 2`] = `
{
  "assignment_type": null,
  "blooms": null,
  "books": [
    {
      "lo": [
        "00-00-01",
      ],
      "name": "stax-psy",
    },
    {
      "aplo": [
        "ABC.0.F",
      ],
      "name": "stax-apbio",
      "science_practice": "concept-explanation",
    },
  ],
  "collaborator_solutions": [],
  "dok": null,
  "errata_id": "test",
  "feature_id": "term-03",
  "feature_page": "m00003",
  "is_solution_public": true,
  "nickname": "new",
  "time": null,
}
`;
